### PR TITLE
feat: delete unused support bundle after a period of time

### PR DESCRIPTION
**Pull request by Yu-Jack**
**Problem:**

Support bundle CR isn't deleted due to refreshing window or something which terminates the downloading process. The leftover support bundle manager occupies the cluster resource.

**Solution:**

***We choose 30 minutes in the end***

Check the exist time with `SupportBundleInitialized` status (This status is in final state).

if this time is longer than 10 minutes, that means it might be broken or error, then delete the support bundle CR. 

I think 10 minutes is enough to wait after the support bundle is ready for downloading.

**Related Issue:**

https://github.com/harvester/harvester/issues/4922

**Test plan:**

1. Create support bundle in harvester dashboard.
2. Refresh the window to avoid the harvester dashboard to trigger download API.
3. Support bundle CR should be deleted after x seconds.
  Go to settings to change `support-bundle-expiration` to like 1 (mean 1 minute), support bundle will be delete in 1 minute. Then, use `k get supportbundles -n harvester-system` to check resource.

### CODE REVIEW

File: pkg/controller/master/supportbundle/controller.go [hunk 1]
```diff
@@ -19,6 +20,8 @@ import (
 	"github.com/harvester/harvester/pkg/settings"
 )
 
+const supportBundleExistTimeLimit = 10 * time.Minute
```
Comment (bk201) [inline pkg/controller/master/supportbundle/controller.go]: Can we make the value longer to 30 minutes and configurable? We can create a setting for this.

Comment (Yu-Jack) [reply pkg/controller/master/supportbundle/controller.go]: surely, sounds good!

File: pkg/settings/settings.go [hunk 2]
```diff
@@ -154,6 +157,10 @@ func (s Setting) GetInt() int {
 	return i
 }
 
+func (s Setting) GetDuration() time.Duration {
+	return time.Duration(s.GetInt())
+}
```
Comment (Yu-Jack) [inline pkg/settings/settings.go]: Just add helper function to convert

File: pkg/settings/settings.go [hunk 3]
```diff
@@ -78,6 +80,7 @@ const (
 	HarvesterCSICCMSettingName                        = "harvester-csi-ccm-versions"
 	StorageNetworkName                                = "storage-network"
 	DefaultVMTerminationGracePeriodSecondsSettingName = "default-vm-termination-grace-period-seconds"
+	SupportBundleExistTimeLimitSecondsSettingName     = "support-bundle-exist-time-seconds"
```
Comment (bk201) [inline pkg/settings/settings.go]: How about `support-bundle-gc-timeout`? Thanks @ibrokethecloud

Comment (Yu-Jack) [reply pkg/settings/settings.go:83]: How about `support-bundle-expired-time` ? `gc-timeout` sounds like it encounters timeout while gcing.

Comment (Yu-Jack) [reply pkg/settings/settings.go]: @bk201 @ibrokethecloud How about `support-bundle-expired-time` ? `gc-timeout` sounds like it encounters timeout while gcing.

Comment (ibrokethecloud) [reply pkg/settings/settings.go]: `support-bundle-expiration-time` ?

Comment (Yu-Jack) [reply pkg/settings/settings.go]: I think `support-bundle-expiration-seconds` would be nice

**Review by bk201 — APPROVED** [review]
LGTM, thanks!

File: pkg/settings/settings.go [hunk 4]
```diff
@@ -154,6 +157,10 @@ func (s Setting) GetInt() int {
 	return i
 }
 
+func (s Setting) GetDuration() time.Duration {
+	return time.Duration(s.GetInt())
```
Comment (w13915984028) [inline pkg/settings/settings.go]: If the setting is cleared, s.GetInt() returns 0, will this function return the default value? And, this value should have a minimal value, user can't set it as 1, 2 ...; for the max value, a limitation seems also required.

BTW, with a new `setting`, the Harvester document needs to updated, do we have another PR for it ? thanks.

File: pkg/controller/master/supportbundle/controller.go [hunk 5]
```diff
@@ -70,12 +71,40 @@ func (h *Handler) OnSupportBundleChanged(key string, sb *harvesterv1.SupportBund
...
+	existTime := time.Now().Sub(t)
+	expiredTime := settings.SupportBundleExpirationSeconds.GetDuration() * time.Second
+
+	logrus.Debugf("[%s] support bundle status: %s exist time is %s", sb.Name, sb.Status.State, existTime.String())
+	if existTime < expiredTime {
+		h.supportBundleController.EnqueueAfter(sb.Namespace, sb.Name, expiredTime)
+		return sb, err
+	}
+
+	if err := h.supportBundles.Delete(sb.Namespace, sb.Name, &metav1.DeleteOptions{}); err != nil {
```
Comment (w13915984028) [inline pkg/controller/master/supportbundle/controller.go]: Skip the error like `NotFound`

File: pkg/controller/master/supportbundle/controller.go [hunk 6]
```diff
@@ -70,12 +71,40 @@ func (h *Handler) OnSupportBundleChanged(key string, sb *harvesterv1.SupportBund
...
 	default:
 		logrus.Debugf("[%s] noop for state %s", sb.Name, sb.Status.State)
 		return sb, nil
 	}
 }
 
+// checkExistTime checks if the support bundle has been updated for more than supportBundleExistTimeLimit minutes
+// If yes, that means the support bundle file is not be retrieved, and it should be deleted.
+func (h *Handler) checkExistTime(sb *harvesterv1.SupportBundle) (*harvesterv1.SupportBundle, error) {
+	t, err := time.Parse(time.RFC3339, sb.Status.Conditions[0].LastUpdateTime)
```
Comment (w13915984028) [inline pkg/controller/master/supportbundle/controller.go]: Here and the test function, assume that the Conditions[0] is the target, it is better to use the provided function to get the target condition, when get, do check, if not, log something and return.

**Review by w13915984028 — COMMENTED** [review]
Please check those robustness related questions, thanks.

File: pkg/settings/settings.go (hunk 4 above)
Comment (Yu-Jack) [reply pkg/settings/settings.go]: Currently, `GetInt` is implemented like below.

```go
func (s Setting) GetInt() int {
	v := s.Get()
	i, err := strconv.Atoi(v)
	if err == nil {
		return i
	}
	logrus.Errorf("failed to parse setting %s=%s as int: %v", s.Name, v, err)
	i, err = strconv.Atoi(s.Default)
	if err != nil {
		return 0
	}
	return i
}
```

So, if the setting is cleared, `GetInt` returns default value and `GetDuration` also return default value. 

It seems there is no any existing mechanism to add this restriction of minimum and maximum, but I think it's good to add it. Maybe we could open another enhancement issue to track this? Because I think it would be better to check other settings's minimum and maximum if it's necessary.

Regarding document, thanks for reminding. I'll check this with @bk201.

Comment (w13915984028) [reply pkg/settings/settings.go]: The value limitation is more checked in webhook.
Assume someone test it with value 1, it will only bring issues to us.

Comment (w13915984028) [reply pkg/settings/settings.go]:
```
	v := s.Get()
	i, err := strconv.Atoi(v)
```

Check `v == ""` first; otherwise, it will print the error

File: pkg/settings/settings.go [hunk 7]
```diff
@@ -141,19 +144,30 @@ func (s Setting) Get() string {
...
+		if i, err = strconv.Atoi(v); err == nil {
+			return i
+		}
+		logrus.Errorf("failed to parse setting %s=%s as int: %v", s.Name, v, err)
 	}
-	logrus.Errorf("failed to parse setting %s=%s as int: %v", s.Name, v, err)
-	i, err = strconv.Atoi(s.Default)
-	if err != nil {
+
+	if i, err = strconv.Atoi(s.Default); err != nil {
```
Comment (w13915984028) [inline pkg/settings/settings.go:161]: return the default value 1800 instead of 0, if this unlucky failure happens

Comment (Yu-Jack) [reply pkg/settings/settings.go:161]: if I failed to parse the default value, how do I return default value? 

If it could parse it correctly, it just return default value, but if it couldn't. The default value might be `1800ss`, I could't also return it as well.

BTW, I've added the webhook to validate the value here.

Comment (Yu-Jack) [reply pkg/settings/settings.go:160]: Or do you mean if I get the zero in controller, I just reassign 1800 to it?

Comment (Yu-Jack) [reply pkg/settings/settings.go:161]: Or do you mean if I get zero in controller, I just reassign 1800 to it?

Comment (w13915984028) [reply pkg/settings/settings.go:161]: This function, in any abnormal case, should return the none-zero default value, at the moment it is 1800. The `0` means the file will be deleted right after it is generated.

This odd situation is caused by below line, the default value can only be a string, and you set the 1800 a direct string.

Maybe you can define 2 consts in the uitl, "1800" as a string and referred in below, and another 1800 as an int, which is used as the fallback value `GetInt() int`

`SupportBundleExpirationSeconds = NewSetting(SupportBundleExpirationSecondsSettingName, "1800")`

Those 2 consts are put together, and with a comment saying they need to changed at mean time.

File: pkg/settings/settings.go [hunk 8]
```diff
@@ -37,6 +38,7 @@ var (
 	SupportBundleImage                     = NewSetting(SupportBundleImageName, "{}")
 	SupportBundleNamespaces                = NewSetting("support-bundle-namespaces", "")
 	SupportBundleTimeout                   = NewSetting(SupportBundleTimeoutSettingName, "10") // Unit is minute. 0 means disable timeout.
+	SupportBundleExpirationSeconds         = NewSetting(SupportBundleExpirationSecondsSettingName, "1800")
```
Comment (w13915984028) [inline pkg/settings/settings.go]:
```
	SupportBundleTimeout                   = NewSetting(SupportBundleTimeoutSettingName, "10") // Unit is minute. 0 means disable timeout.
	SupportBundleExpirationSeconds         = NewSetting(SupportBundleExpirationSecondsSettingName, "1800")
```

Another question: what's the consideration to use different time units here ?

Comment (Yu-Jack) [reply pkg/settings/settings.go]: Different usage.

- SupportBundleTimeout
  It is used to check if support bundle manager execute too long, such as collecting logs. It might take some times if you have a lots of nodes.
- SupportBundleExpirationSeconds 
  It is used to remove useless support bundle CR after finished collecting logs.

File: pkg/settings/settings.go (hunk 7 above)
Comment (Yu-Jack) [reply pkg/settings/settings.go:161]: I think it's weird to put two const here. I think we could change the `setting.Def` type into `interface{}`. Then you could put anything you want instead of putting two const due to the constraint of `setting.Def`.

File: pkg/settings/settings.go (hunk 8 above)
Comment (w13915984028) [reply pkg/settings/settings.go]: Seems no clear reason that the first setting is in `minutes` but the second is in `seconds`, it may confuse the users.

Could `SupportBundleExpiration` also use minutes ?

Comment (Yu-Jack) [reply pkg/settings/settings.go]: Surely, then I will remove `Seconds` word.

File: pkg/settings/settings.go (hunk 7 above)
Comment (w13915984028) [reply pkg/settings/settings.go:161]:
```
if i, err = strconv.Atoi(s.Default); err != nil {
  return 30 // minutes
}
```

should be fine in this case

Comment (Yu-Jack) [reply pkg/settings/settings.go:161]: If I understand correctly, it will affect other setting, right?

Comment (Yu-Jack) [reply pkg/settings/settings.go:161]: I think we just keep it returning 0 instead of returning 30 here. Otherwise, it affects other setting and makes some confusion about why we put 30 just only for support-bundle-expiration.

Comment (w13915984028) [reply pkg/settings/settings.go:161]: Indeed, 0 is an invalid value for this function, it should never be returned.

In any case, when return 0, the generated SB file will be removed right away, this breaks the feature.

Different settings have different default value and value range. This setting, should be at least e.g. 2 minutes, thus the SB file will have chance to be downloaded.

Recall the SB feature, the priority is:
(1) The SB file can be generated or fail
(2) The SB file can be downloaded or fail
(3) The SB file can be deleted after certain time.

Comment (Yu-Jack) [reply pkg/settings/settings.go:161]: But, if you return 30 in this function, it also break other features. For example, [support-bundle-timeout](https://docs.harvesterhci.io/v1.2/advanced/index#support-bundle-timeout) uses 0 as disabling. So, I suggest we need to enhance whole setting structure, otherwise it would be weird case here, like you said 0 is an invalid value. I think it's a different issue here. It's kind of existing setting struct issue.

Comment (w13915984028) [reply pkg/settings/settings.go:161]: OK, then please replace 0 in the controller side, revert 0 to 30.

Comment (Yu-Jack) [reply pkg/settings/settings.go:161]: Surely, it makes sense for me.

File: pkg/util/supportbundle/constants.go [hunk 9]
```diff
@@ -0,0 +1,7 @@
+package supportbundle
+
+const (
+	// SupportBundleExpirationDefault and SupportBundleExpirationDefaultStr these two value should be changed together.
+	SupportBundleExpirationDefault    = 30
+	SupportBundleExpirationDefaultStr = "30"
+)
```
Comment (Yu-Jack) [inline pkg/util/supportbundle/constants.go:7]: Due to import cycles, so I move out constants to another folder under utils. These two value should be removed after redesign the setting structure in the future.

**Review by w13915984028 — APPROVED** [review]
LGTM, thanks.

btw, please squash the commits before final merge.

File: pkg/controller/master/supportbundle/controller.go [hunk 10]
```diff
@@ -70,12 +73,48 @@ func (h *Handler) OnSupportBundleChanged(key string, sb *harvesterv1.SupportBund
...
 		return sb, nil
 	}
 }
 
+// checkExistTime checks if the support bundle has been updated for more than supportBundleExistTimeLimit minutes
+// If yes, that means the support bundle file is not be retrieved, and it should be deleted.
+func (h *Handler) checkExistTime(sb *harvesterv1.SupportBundle) (*harvesterv1.SupportBundle, error) {
+	t, err := time.Parse(time.RFC3339, harvesterv1.SupportBundleInitialized.GetLastUpdated(sb))
+	if err != nil {
+		logrus.Debugf("[%s] fail to parse %s", sb.Name, sb.Status.Conditions[0].LastUpdateTime)
```
Comment (ibrokethecloud) [inline pkg/controller/master/supportbundle/controller.go]: we should probably use `harvesterv1.SupportBundleInitialized.GetLastUpdated(sb)` in the error message, rather than assuming this will always be the first condition. This should ensure that we future proof against any more conditions being added.

File: pkg/controller/master/supportbundle/controller.go [hunk 11]
```diff
@@ -70,12 +73,48 @@ func (h *Handler) OnSupportBundleChanged(key string, sb *harvesterv1.SupportBund
...
+	)
+
+	if expiration := settings.SupportBundleExpiration.GetInt(); expiration == 0 {
+		expiredTime = time.Duration(supportBundleUtil.SupportBundleExpirationDefault) * time.Minute
+	} else {
+		expiredTime = time.Duration(expiration) * time.Minute
+	}
+
+	logrus.Debugf("[%s] support bundle status: %s exist time is %s", sb.Name, sb.Status.State, existTime.String())
+	if existTime < expiredTime {
```
Comment (ibrokethecloud) [inline pkg/controller/master/supportbundle/controller.go:105]: minor nit, we always requeue after `expiredTime` which will be the expiry duration coming from the setting. If we want to gc the crd earlier we should probably requeue after `expiredTime - existTime`

File: pkg/controller/master/supportbundle/controller.go (hunk 10 above)
Comment (Yu-Jack) [reply pkg/controller/master/supportbundle/controller.go]: I forgot to modify usage in debug message.

**Review by ibrokethecloud — APPROVED** [review]
lgtm. thanks.
